### PR TITLE
Add ability to exclude tests for deprecated endpoints

### PR DESCRIPTION
**Pull request by neilalexander**
This adds a new `deprecated_endpoints` qw to `require`, and adds a command-line flag `--exclude-deprecated`. The idea is that we can not bother running tests at all that test endpoints that we know are deprecated, e.g. in Dendrite we will never implement `/initialSync`.

### CODE REVIEW

File: run-tests.pl [hunk 1]
```diff
@@ -899,6 +919,7 @@ sub list_symbols
       passed  => $passed_count,
       failed  => $failed_count,
       skipped => $skipped_count,
+	  excluded => $excluded_count,
```
Comment (clokep) [inline run-tests.pl:922]:
```suggestion
      excluded => $excluded_count,
```

File: run-tests.pl [hunk 2]
```diff
@@ -177,7 +182,10 @@ sub usage
    -n, --no-tls                 - prefer plaintext client connections where
                                   possible
 
-       --bind-host HOST         - when starting listeners (eg homeservers or
```
Comment (richvdh) [inline run-tests.pl:180]: this was deliberate. the `--long-options` all line up.

Comment (neilalexander) [reply run-tests.pl:180]: Oh whoops. I'll fix.

File: run-tests.pl [hunk 3]
```diff
@@ -838,24 +846,36 @@ sub list_symbols
...
    my $m = $test->multi ? "enter_multi_test" : "enter_test";
 
+   # If the test specifically tests deprecated endpoints and we're not allowing those
+   # then skip the test altogether - there's no point in even showing it as skipped.
+   my @requires = @{ $test->requires // [] };
+   foreach my $req ( @requires ) {
+		if ($req eq "deprecated_endpoints" && !$INCLUDE_DEPRECATED_ENDPOINTS) {
+		   $excluded_count++;
+		   next TESTS;
+		}
```
Comment (richvdh) [inline run-tests.pl:876]: indentation

File: tests/30rooms/01state.pl [hunk 4]
```diff
@@ -287,7 +288,8 @@ sub set_test_state
    };
 
 test "Joining room twice is idempotent",
-   requires => [ local_user_and_room_fixtures() ],
+   # TODO: deprecated endpoint used in this test
+   requires => [ local_user_and_room_fixtures()],
```
Comment (richvdh) [inline tests/30rooms/01state.pl:292]:
```suggestion
   requires => [ local_user_and_room_fixtures() ],
```

File: tests/21presence-events.pl [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 test "initialSync sees my presence status",
    requires => [ local_user_fixture( with_events => 1 ),
-                 qw( can_initial_sync )],
+                 qw( can_initial_sync, deprecated_endpoints )],
```
Comment (richvdh) [inline tests/21presence-events.pl:7]:
```suggestion
                 qw( can_initial_sync deprecated_endpoints )],
```

File: tests/30rooms/30history-visibility.pl [hunk 6]
```diff
@@ -131,7 +131,7 @@ sub test_history_visibility
    test(
       "$name non-joined user can call /events on world_readable room",
 
-      requires => [ $fixture->(), local_user_fixture( with_events => 1 ), local_user_fixture( with_events => 1 ) ],
+      requires => [ $fixture->(), local_user_fixture( with_events => 1 ), local_user_fixture( with_events => 1 ), qw ( deprecated_endpoints ) ],
```
Comment (richvdh) [inline tests/30rooms/30history-visibility.pl:134]: long line is long

```suggestion
      requires => [ 
         $fixture->(), local_user_fixture( with_events => 1 ), local_user_fixture( with_events => 1 ), 
         qw ( deprecated_endpoints ),
      ],
```

**Pull request merged**